L1Trigger/VertexFinder/plugins/VertexAnalyzer.cc

```diff
@@ -0,0 +1,820 @@
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "L1Trigger/VertexFinder/interface/AnalysisSettings.h"
+#include "L1Trigger/VertexFinder/interface/InputData.h"
+#include "L1Trigger/VertexFinder/interface/L1TrackTruthMatched.h"
+#include "L1Trigger/VertexFinder/interface/RecoVertex.h"
+#include "L1Trigger/VertexFinder/interface/selection.h"
+#include "SimDataFormats/GeneratorProducts/interface/HepMCProduct.h"
+#include "SimDataFormats/TrackingAnalysis/interface/TrackingParticle.h"
+#include "SimTracker/TrackTriggerAssociation/interface/TTClusterAssociationMap.h"
+#include "SimTracker/TrackTriggerAssociation/interface/TTStubAssociationMap.h"
+
+#include "TEfficiency.h"
+#include "TGraphAsymmErrors.h"
+#include "TGraphErrors.h"
+#include "TH1F.h"
+#include "TH2F.h"
+#include "TPad.h"
+#include "TProfile.h"
+
+#include <map>
+#include <string>
+#include <vector>
+#include <iostream>
+
+using namespace std;
+
+namespace l1tVertexFinder {
+
+  class VertexAnalyzer : public edm::EDAnalyzer {
+  public:
+    explicit VertexAnalyzer(const edm::ParameterSet&);
+    ~VertexAnalyzer() override;
+
+  private:
+    template <typename T>
+    int getIndex(std::vector<T> collection, T reference);
+    void beginJob() override;
+    void analyze(const edm::Event& evt, const edm::EventSetup& setup) override;
+    void endJob() override;
+
+    // define types for stub-related classes
+    typedef edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>> DetSetVec;
+    typedef TTTrackAssociationMap<Ref_Phase2TrackerDigi_> TTTrackAssMap;
+    typedef TTStubAssociationMap<Ref_Phase2TrackerDigi_> TTStubAssMap;
+    typedef TTClusterAssociationMap<Ref_Phase2TrackerDigi_> TTClusterAssMap;
+    typedef edm::View<TTTrack<Ref_Phase2TrackerDigi_>> TTTrackCollectionView;
+
+    // references to tags containing information relevant to perofrmance analysis
+    const edm::EDGetTokenT<edm::HepMCProduct> hepMCInputTag;
+    const edm::EDGetTokenT<edm::View<reco::GenParticle>> genParticleInputTag;
+    const edm::EDGetTokenT<TrackingParticleCollection> tpInputTag;
+    const edm::EDGetTokenT<TTTrackAssMap> trackTruthInputTag;
+    const edm::EDGetTokenT<DetSetVec> stubInputTag;
+    const edm::EDGetTokenT<TTStubAssMap> stubTruthInputTag;
+    const edm::EDGetTokenT<TTClusterAssMap> clusterTruthInputTag;
+    const edm::EDGetTokenT<TTTrackCollectionView> l1TracksToken_;
+    const edm::EDGetTokenT<std::vector<l1t::Vertex>> l1VerticesToken_;
+
+    const bool printResults_;
+
+    // storage class for configuration parameters
+    AnalysisSettings settings_;
+
+    edm::Service<TFileService> fs_;
```

Comment:
Service handles should not be stored as class member variables. Instead, instantiate it in each member function it is needed.